#include "includes.hpp"
#include "ui_stageconfigeditorv5.h"
#include "stageconfigeditorv5.hpp"
#include "objectselector.hpp"

#include <RSDKv5/stageconfigv5.hpp>

StageConfigEditorv5::StageConfigEditorv5(RSDKv5::StageConfig *scf, QList<SceneObject> &objList, QList<GameObjectInfo> gamelinkObjs, bool linkState, QWidget *parent)
    : QDialog(parent), ui(new Ui::StageConfigEditorv5), stageConfig(scf)
{
    ui->setupUi(this);

    this->setWindowTitle("StageConfig Editor");
    // remove question mark from the title bar
    setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);

    int gcCount = objList.count() - stageConfig->objects.count();
    linkLoaded = linkState;
    setupUI();

    ui->sectionList->blockSignals(true);
    ui->sectionList->setCurrentRow(-1);
    ui->sectionList->blockSignals(false);
    ui->sectionList->setCurrentRow(0);

    connect(ui->sectionList, &QListWidget::currentRowChanged, [this](int v) {
        ui->objList->blockSignals(true);
        ui->objList->setCurrentRow(0);
        ui->objList->blockSignals(false);
        ui->objList->setCurrentRow(-1);

        ui->sfxList->blockSignals(true);
        ui->sfxList->setCurrentRow(0);
        ui->sfxList->blockSignals(false);
        ui->sfxList->setCurrentRow(-1);

        ui->detailsWidget->setCurrentIndex(v);
        ui->sectionLabel->setText(ui->sectionList->item(v)->text());
    });

    // ----------------
    // OBJECTS
    // ----------------
    connect(ui->objList, &QListWidget::currentRowChanged, [this](int c) {
        if (ui->upObj)
            ui->upObj->setDisabled(c == -1);
        if (ui->downObj)
            ui->downObj->setDisabled(c == -1);
        if (ui->rmObj)
            ui->rmObj->setDisabled(c == -1);

        ui->objName->setDisabled(c == -1);

        ui->replaceObj->setDisabled(!linkLoaded || c == -1);

        if (c == -1)
            return;

        ui->objName->blockSignals(true);
        ui->objName->setText(stageConfig->objects[c]);
        ui->objName->setEnabled(!linkLoaded);
        ui->objName->blockSignals(false);

        if (ui->downObj)
            ui->downObj->setDisabled(c == ui->objList->count() - 1);
        if (ui->upObj)
            ui->upObj->setDisabled(c == 0);
    });

    connect(ui->addObj, &QToolButton::clicked, [this, gcCount, &objList, gamelinkObjs] {
        uint c = ui->objList->currentRow() + 1;
        if (!linkLoaded){
            stageConfig->objects.insert(c, "New Object");
            auto *item = new QListWidgetItem();
            item->setText(stageConfig->objects[c]);
            ui->objList->insertItem(c, item);

            item->setFlags(item->flags() | Qt::ItemIsEditable);
            ui->objList->setCurrentItem(item);
        } else{
            QList<QString> nameList;
            for (int s = 0; s < gcCount; s++){
                nameList.append(objList[s].name);
            }
            for (int s = 0; s < ui->objList->count(); s++){
                nameList.append(ui->objList->item(s)->text());
            }

            ObjectSelectorv5 *selector = new ObjectSelectorv5(nameList, gamelinkObjs);
            if (selector->exec() == QDialog::Accepted) {
                for (int i = selector->objAddList.count() - 1; i >= 0; --i) {
                    if (selector->objAddList[i]) {
                        stageConfig->objects.insert(c, gamelinkObjs[selector->objIDList[i]].name);

                        auto *item = new QListWidgetItem();
                        item->setText(stageConfig->objects[c]);
                        ui->objList->insertItem(c, item);
                    }
                }

            }
        }
        DoAction("Added Object");
    });

    connect(ui->upObj, &QToolButton::clicked, [this] {
        uint c     = ui->objList->currentRow();
        auto *item = ui->objList->takeItem(c);
        stageConfig->objects.move(c, c - 1);
        ui->objList->insertItem(c - 1, item);
        ui->objList->setCurrentRow(c - 1);
        DoAction("Moved Object Up");
    });

    connect(ui->downObj, &QToolButton::clicked, [this] {
        uint c     = ui->objList->currentRow();
        auto *item = ui->objList->takeItem(c);
        stageConfig->objects.move(c, c + 1);
        ui->objList->insertItem(c + 1, item);
        ui->objList->setCurrentRow(c + 1);
        DoAction("Moved Object Down");
    });

    connect(ui->rmObj, &QToolButton::clicked, [this] {
        int c = ui->objList->currentRow();
        int n = ui->objList->currentRow() == ui->objList->count() - 1 ? c - 1 : c;
        delete ui->objList->item(c);
        stageConfig->objects.removeAt(c);
        ui->objList->blockSignals(true);
        ui->objList->setCurrentRow(n);
        ui->objList->blockSignals(false);
        DoAction("Removed Object");
    });

    connect(ui->replaceObj, &QToolButton::clicked, [this, gcCount, &objList, gamelinkObjs] {
        oldObj = ui->objList->currentItem()->text();
        QList<QString> nameList;
        for (int s = 0; s < gcCount; s++){
            nameList.append(objList[s].name);
        }
        for (int s = 0; s < ui->objList->count(); s++){
            nameList.append(ui->objList->item(s)->text());
        }

        ObjectSelectorv5 *replacer = new ObjectSelectorv5(nameList, gamelinkObjs, true);
        if (replacer->exec() == QDialog::Accepted) {
            ui->objList->currentItem()->setText(replacer->replacedObj);
            stageConfig->objects[ui->objList->currentRow()] = replacer->replacedObj;

            int replacedID = 0;
            for (int i = 0; i < objList.count(); i++){
                if (objList[i].name == oldObj){
                    replacedID = i;
                    break;
                }
            }

            for (int o = 0; o < gamelinkObjs.count(); ++o) {
                if (gamelinkObjs[o].name == replacer->replacedObj){
                    SceneObject obj;
                    obj.name = replacer->replacedObj;
                    objList[replacedID].name = replacer->replacedObj;
                    break;
                }
            }
        }
        DoAction("Replaced Object");
    });


    connect(ui->objList, &QListWidget::itemChanged, [this](QListWidgetItem *item) {
        stageConfig->objects[ui->objList->row(item)] = item->text();

        ui->objName->blockSignals(true);
        ui->objName->setText(stageConfig->objects[ui->objList->row(item)]);
        ui->objName->blockSignals(false);

        DoAction("Changed Object Name");
    });

    connect(ui->objName, &QLineEdit::textEdited, [this](QString s) {
        stageConfig->objects[ui->objList->currentRow()] = s;

        ui->objList->blockSignals(true);
        ui->objList->item(ui->objList->currentRow())
            ->setText(stageConfig->objects[ui->objList->currentRow()]);
        ui->objList->blockSignals(false);

        DoAction("Changed Object Name");
    });

    // ----------------
    // SOUNDFX
    // ----------------
    connect(ui->sfxList, &QListWidget::currentRowChanged, [this](int c) {
        if (ui->upSfx)
            ui->upSfx->setDisabled(c == -1);
        if (ui->downSfx)
            ui->downSfx->setDisabled(c == -1);
        if (ui->rmSfx)
            ui->rmSfx->setDisabled(c == -1);

        ui->sfxPath->setDisabled(c == -1);
        ui->maxPlays->setDisabled(c == -1);

        if (c == -1)
            return;

        ui->sfxPath->blockSignals(true);
        ui->sfxPath->setText(stageConfig->soundFX[c].path);
        ui->sfxPath->blockSignals(false);

        ui->maxPlays->blockSignals(true);
        ui->maxPlays->setValue(stageConfig->soundFX[c].maxConcurrentPlay);
        ui->maxPlays->blockSignals(false);

        if (ui->downSfx)
            ui->downSfx->setDisabled(c == ui->sfxList->count() - 1);
        if (ui->upSfx)
            ui->upSfx->setDisabled(c == 0);
    });

    connect(ui->addSfx, &QToolButton::clicked, [this] {
        uint c = ui->sfxList->currentRow() + 1;
        stageConfig->soundFX.insert(c, RSDKv5::StageConfig::WAVConfiguration());
        auto *item = new QListWidgetItem();
        item->setText(stageConfig->soundFX[c].path);
        ui->sfxList->insertItem(c, item);

        item->setFlags(item->flags() | Qt::ItemIsEditable);
        ui->sfxList->setCurrentItem(item);
        DoAction("Added Sfx");
    });

    connect(ui->upSfx, &QToolButton::clicked, [this] {
        uint c     = ui->sfxList->currentRow();
        auto *item = ui->sfxList->takeItem(c);
        stageConfig->soundFX.move(c, c - 1);
        ui->sfxList->insertItem(c - 1, item);
        ui->sfxList->setCurrentRow(c - 1);
        DoAction("Moved Sfx Up");
    });

    connect(ui->downSfx, &QToolButton::clicked, [this] {
        uint c     = ui->sfxList->currentRow();
        auto *item = ui->sfxList->takeItem(c);
        stageConfig->soundFX.move(c, c + 1);
        ui->sfxList->insertItem(c + 1, item);
        ui->sfxList->setCurrentRow(c + 1);
        DoAction("Moved Sfx Down");
    });

    connect(ui->rmSfx, &QToolButton::clicked, [this] {
        int c = ui->sfxList->currentRow();
        int n = ui->sfxList->currentRow() == ui->sfxList->count() - 1 ? c - 1 : c;
        delete ui->sfxList->item(c);
        stageConfig->soundFX.removeAt(c);
        ui->sfxList->blockSignals(true);
        ui->sfxList->setCurrentRow(n);
        ui->sfxList->blockSignals(false);
        DoAction("Removed Sfx");
    });

    connect(ui->sfxList, &QListWidget::itemChanged, [this](QListWidgetItem *item) {
        stageConfig->soundFX[ui->sfxList->row(item)].path = item->text();

        ui->sfxPath->blockSignals(true);
        ui->sfxPath->setText(stageConfig->soundFX[ui->sfxList->row(item)].path);
        ui->sfxPath->blockSignals(false);
        DoAction("Changed Sfx Path");
    });

    connect(ui->sfxPath, &QLineEdit::textEdited, [this](QString s) {
        stageConfig->soundFX[ui->sfxList->currentRow()].path = s;
        DoAction("Changed Sfx Path");

        ui->sfxList->blockSignals(true);
        ui->sfxList->item(ui->sfxList->currentRow())
            ->setText(stageConfig->soundFX[ui->sfxList->currentRow()].path);
        ui->sfxList->blockSignals(false);
    });

    connect(ui->maxPlays, QOverload<int>::of(&QSpinBox::valueChanged), [this](int v) {
        stageConfig->soundFX[ui->sfxList->currentRow()].maxConcurrentPlay = v;

        ui->sfxList->blockSignals(true);
        ui->sfxList->item(ui->sfxList->currentRow())
            ->setText(stageConfig->soundFX[ui->sfxList->currentRow()].path);
        ui->sfxList->blockSignals(false);

        DoAction("Changed Sfx Plays");
    });
}

StageConfigEditorv5::~StageConfigEditorv5() { delete ui; }

bool StageConfigEditorv5::event(QEvent *event)
{

    switch ((int)event->type()) {
        default: break;
        case RE_EVENT_UNDO: UndoAction(); break;
        case RE_EVENT_REDO: RedoAction(); break;
    }

    return QWidget::event(event);
}

void StageConfigEditorv5::setupUI(bool allowRowChange)
{
    if (allowRowChange) {
        ui->sectionList->blockSignals(true);
        ui->sectionList->setCurrentRow(-1);
        ui->sectionList->blockSignals(false);
        ui->sectionList->setCurrentRow(0);
    }

    ui->objList->blockSignals(true);
    ui->sfxList->blockSignals(true);

    int prevIndex = ui->objList->currentRow();
    ui->objList->setCurrentRow(0);
    ui->objList->setCurrentRow(prevIndex >= ui->objList->count() ? -1 : prevIndex);

    prevIndex = ui->sfxList->currentRow();
    ui->sfxList->setCurrentRow(0);
    ui->sfxList->setCurrentRow(prevIndex >= ui->sfxList->count() ? -1 : prevIndex);

    ui->objList->clear();
    int id = 0;
    for (QString &obj : stageConfig->objects) {
        ui->objList->addItem(obj);
        if (!linkLoaded)
            ui->objList->item(id)->setFlags(ui->objList->item(id)->flags() | Qt::ItemIsEditable);
        id++;
    }

    ui->sfxList->clear();
    id = 0;
    for (auto &sfx : stageConfig->soundFX) {
        ui->sfxList->addItem(sfx.path);
        ui->sfxList->item(id)->setFlags(ui->sfxList->item(id)->flags() | Qt::ItemIsEditable);
        id++;
    }

    ui->objList->blockSignals(false);
    ui->sfxList->blockSignals(false);
}

void StageConfigEditorv5::UndoAction()
{
    if (actionIndex > 0) {
        // setStatus("Undid Action: " + actions[actionIndex].name);
        actionIndex--;
        ResetAction();
    }
}
void StageConfigEditorv5::RedoAction()
{
    if (actionIndex + 1 < actions.count()) {
        // setStatus("Redid Action: " + actions[actionIndex].name);
        actionIndex++;
        ResetAction();
    }
}
void StageConfigEditorv5::ResetAction()
{
#if RE_USE_UNSTABLE
    copyConfig(NULL, &actions[actionIndex]);

    setupUI(false);
#endif
}
void StageConfigEditorv5::DoAction(QString name, bool setModified)
{
    Q_UNUSED(setModified);

    ActionState action;

    action.name = name;

#if RE_USE_UNSTABLE
    copyConfig(&action, NULL);
#endif

    // Actions
    for (int i = actions.count() - 1; i > actionIndex; --i) {
        actions.removeAt(i);
    }

    actions.append(action);
    actionIndex = actions.count() - 1;

    // setStatus("Did Action: " + name);
}
void StageConfigEditorv5::ClearActions()
{
    actions.clear();
    actionIndex = 0;
    DoAction("Action Setup", false); // first action, cant be undone
}

void StageConfigEditorv5::copyConfig(ActionState *stateDst, ActionState *stateSrc)
{
    RSDKv5::StageConfig *src = stateSrc ? &stateSrc->stageConfig : stageConfig;
    RSDKv5::StageConfig *dst = stateDst ? &stateDst->stageConfig : stageConfig;

    dst->filePath          = src->filePath;
    dst->loadGlobalObjects = src->loadGlobalObjects;

    for (int i = 0; i < 8; ++i) {
        for (int r = 0; r < 16; ++r) {
            dst->palettes[i].activeRows[r] = src->palettes[i].activeRows[r];
            for (int c = 0; c < 16; ++c) dst->palettes[i].colors[r][c] = src->palettes[i].colors[r][c];
        }
    }

    dst->objects.clear();
    for (auto obj : src->objects) dst->objects.append(obj);

    dst->soundFX.clear();
    for (auto sfx : src->soundFX) dst->soundFX.append(sfx);
}

#include "moc_stageconfigeditorv5.cpp"
